Only use prebuilt binaries when they are compatible with the local version of Swift

scottrhoyt opened the pull request:
Closes #1354

This is a first stab at implementing Swift version checking for prebuilt binary frameworks. The current approach is to compare the output of `xcrun swift --version` with the Swift version information at the top of the `Framework-Swift.h` file in Swift frameworks. The versions must match exactly (via string comparison) to be considered compatible.

The compatibility check is pessimistic, so if for any reason the compatibility status cannot be determined, Carthage will build the project instead.

**TODO:**
- [x] allow objective-c frameworks to bypass Swift compatibility check.

Thanks @neonichu for the pointer on possible solutions.

scottrhoyt commented:
If anyone has any pointers on how to distinguish Objective-C frameworks from Swift or whether that is possible/advisable, let me know.

mdiep commented on Source/CarthageKit/Project.swift:
```diff
@@ -117,6 +118,10 @@ public enum ProjectEvent {
+
+	/// Installing of a binary framework is being skipped because of an inability
+	/// to verify that it was built with a compatible Swift version.
+	case skippedInstallingBinaries(ProjectIdentifier, String)
```
How about `skippedIncompatibleBinary`?

Also what's the `String` here? Can you add a label to compensate for the poor type information?

scottrhoyt replied:
I was trying to make this event more general than just the incompatible swift versions event, incase other reasons for skipping a binary installation were added. I'd love to add the labels, I was just copying the local coding style as suggesting in `CONTRIBUTING`. Will change.

mdiep replied:
> I can create an internal error with these three failure modes and wrap them into a single CarthageError.

That would be great. ✨

mdiep replied:
I don't think they even need a case on `CarthageError`. If you create the internal error, you can pass that to the `ProjectEvent` case.

mdiep commented on Source/CarthageKit/Project.swift:
```diff
@@ -171,6 +181,38 @@ public final class Project {
+					return SignalProducer(error: .unknownLocalSwiftVersion)
+				}
+			}
+	}
```
Can you move this to `Xcode.swift`? It's really an Xcode thing, not Carthage-specific project logic.

mdiep commented on Source/CarthageKit/Project.swift:
```diff
@@ -453,23 +495,22 @@ public final class Project {
-							}
+						.map { _ in true }
+						.flatMapError { error in
+							self._projectEventsObserver.send(value: .skippedInstallingBinaries(project, error.description))
```
It's probably better to just pass `error` to the enum.

scottrhoyt replied:
I followed the pattern on `skippedDownloadingBinaries` here. I do think it would be helpful to include the project information as well, but I suppose I could fold that in to an error.

mdiep commented on Source/CarthageKit/Project.swift:
```diff
@@ -531,6 +572,31 @@ public final class Project {
+		}
+
+		return SignalProducer(value: swiftVersion)
+	}
```
This should probably move to `Xcode.swift` as well.

scottrhoyt replied:
Do you think it makes sense for a `Project` to be able to determine it's framework's Swift version? Should I leave that in here or move it all over to `Xcode.swift`.

mdiep commented on Source/CarthageKit/Project.swift:
```diff
@@ -531,6 +572,31 @@ public final class Project {
+					? SignalProducer(value: frameworkURL)
+					: SignalProducer(error: .incompatibleFrameworkSwiftVersions(local: swiftVersion, framework: frameworkVersion))
+			}
+			.take(first: 1)
```
This `take(first:)` shouldn't be needed.

scottrhoyt replied:
Sorry, that's an old habit when doing a `combineLatest` and intending for the result to be a `Single` since if anything changes with the signals you are observing you could easily wind up producing more than a single value. Because that would mean in this case multiple copies of a framework in this case, I thought I'd be safe. I will remove it.

mdiep commented on Source/carthage/Extensions.swift:
```diff
@@ -112,6 +112,9 @@ internal struct ProjectEventSink {
 			carthage.println(formatting.bullets + "Skipped building " + formatting.projectName(project.name) + " due to the error:\n" + message)
+
+		case let .skippedInstallingBinaries(project, message):
+			carthage.println(formatting.bullets + "Skipped installing " + formatting.projectName(project.name) + ".framework binary due to the error:\n\t" + formatting.quote(message))
```
Maybe this doesn't even need `message` at all.

Maybe we should aim for something like this?

```
*** Skipped installing Result.framework binary built with Swift 2.2
```

scottrhoyt replied:
This is related to my comments on errors above. We can possibly remove the message. It seems like it would be disingenuous to say:

```bash
*** Skipped installing Result.framework binary built with Swift 2.2
```

when in fact, it was skipped because the parsing of the local swift version had failed. Either way, I do think it is very helpful to have the extra verbosity of local and framework swift versions here where possible. I think it's very useful when looking at CI logs--for instance--to figure out if it was the Xcode version change or a `carthage update` that is now causing your build to take much longer.

mdiep replied:
You're right. I think keeping the different cases is a good idea.

scottrhoyt commented on Source/CarthageKit/Errors.swift:
```diff
@@ -86,6 +86,15 @@ public enum CarthageError: Error, Equatable {
+	case unknownFrameworkSwiftVersion
+
+	/// The framework binary is not compatible with the local Swift version.
+	case incompatibleFrameworkSwiftVersions(local: String, framework: String)
```
The idea here was to have them all be user-facing. This regex-based parsing feels a little fragile to me. So since the check is pessimistic, this was intended to give the developer as much information as possible to diagnose why their framework isn't being used. Failing with these three error outputs:

```bash
*** Skipped installing BadOld.framework binary due to the error:
	"Incompatible Swift version - framework was built with 3.0 and the local version is 3.0.2."
```

```bash
*** Skipped installing BadOld.framework binary due to the error:
	"Unable to determine local Swift version."
```

```bash
*** Skipped installing BadOld.framework binary due to the error:
	"Unable to determine framework Swift version."
```

If you think this approach is undesirable, I can either reduce this verbosity down to a single error message, or I can create an internal error with these three failure modes and wrap them into a single `CarthageError`.

scottrhoyt commented:
Appreciate the feedback @mdiep! As for the ObjC issue, one possible solution is to make this check optimistic. It would result in behavior that is, at worst, the same as right now, but it would allow ObjC frameworks through and catch most of the incompatible Swift frameworks.

scottrhoyt commented on Source/CarthageKit/Errors.swift:
```diff
@@ -291,6 +309,15 @@ extension CarthageError: CustomStringConvertible {
+			return "Unable to determine local Swift version."
+
+		case .unknownFrameworkSwiftVersion:
+			return "Unable to determine local Swift version."
```
Need to fix this.

dcaunt commented:
One indicator of a framework containing Swift is the presence of a `.swiftmodule` in the Modules subdirectory (e.g. `CarthageKit.framework/Modules/CarthageKit.swiftmodule`). The absence may suggest an Objective-C framework.

scottrhoyt commented:
Thanks for the pointer @dcaunt! I have found documentation in this area a bit hard to come by. I'll do a little digging into prebuilt Objective-C frameworks and check it out. If anyone has a good set of them, let me know.

scottrhoyt commented:
Ok, I did a little check and it does appear to be the case that pure Objective-C frameworks have no .swiftmodule while mixed language frameworks do. Here are the libraries I looked at:

**Pure ObjC**
* AFNetworking
* OCMock
* Realm
* Mantle

**Mixed**
* RealmSwift
* OHHTTPStubs

That's enough to satisfy me. I will add a check for swiftmodule and use the skip Swift version validation when it isn't present.

scottrhoyt commented:
@mdiep I implemented the strategy outlines above for detecting ObjC frameworks (thanks @dcaunt !). I corrected many of your comments. The only thing left is how to handle `CarthageError` and `ProjectEvent`. For these I propose that we follow the current patterns in the source and then open up a separate PR to refactor them as a whole to be consistent. The last main point is what granularity to provide to the developer as to why we are rejecting a prebuilt binary. I think it is useful to offer this granularity because I don't have confidence that the output of `swift --version` won't ever change and it is still possible to make a Swift framework without the `Framework-Swift.h` file as @neonichu notes. 

What do you think?

scottrhoyt commented:
Interestingly enough, it seems the mainline version of Carthage still builds ObjC frameworks even after downloading (tested with Mantle and AFNetworking). Not quite sure why, but that behavior persists with this feature as well.

mdiep commented on Source/CarthageKit/Xcode.swift:
```diff
@@ -16,6 +16,60 @@ import XCDBLD
+			} else {
+				return SignalProducer(error: .unknownLocalSwiftVersion)
+			}
+	}
```
This `flatMap` can be written like this:

```swift
.attemptMap { Result($0, failWith: CarthageError.unknownLocalSwiftVersion) }
```

mdiep commented:
Once the errors are cleaned up like we discussed above, I'll spend some time testing this out.

This will be a huge help. Thanks for working on this!

scottrhoyt commented:
@mdiep ok, I pulled the errors out into a `SwiftVersionError.swift`. I still needed to create a `CarthageError.internalError` to `mapError` to. At that point it didn't make much sense to keep any of the swift version checking login in `Project`, so I moved that all out to`Xcode.swift`. `ProjectEvent.skippingInstallingBinaries` now takes an `Error` and constructs its message from that.

As an aside, at some point it might make more sense to use the `LocalizeError` protocol instead of `CustomStringConvertible` to communicate error descriptions.

mdiep commented on Source/CarthageKit/Project.swift:
```diff
@@ -117,6 +118,10 @@ public enum ProjectEvent {
+
+	/// Installing of a binary framework is being skipped because of an inability
+	/// to verify that it was built with a compatible Swift version.
+	case skippedInstallingBinaries(project: ProjectIdentifier, error: Error)
```
This seems like it should be a `SwiftVersionError`.

scottrhoyt replied:
See comment on generalized `installBinariesForProject` error handling.

mdiep commented on Source/CarthageKit/Project.swift:
```diff
@@ -453,23 +458,25 @@ public final class Project {
+						.map { _ in true }
+						.flatMapError { error in
+							self._projectEventsObserver.send(value: .skippedInstallingBinaries(project: project, error: error))
+							return SignalProducer(value: false)
```
This `flatMapError` could replace the `mapError` after `checkFrameworkCompatibility`. Then you can get rid of `CarthageError.internalError`.

scottrhoyt replied:
The idea here is to handle any `CarthageError` produced by the `installBinaries` step. This is also why the `skippedInstallingBinaries` project message is generic, so that it can cover a range of errors--like errors in copying the frameworks artifacts. This is similar to the semantics for `skippedDownloadingBinaries` which is a general message that in practice is used only if you aren't authenticated to GitHub.

Also, I don't how we could move around the `flatMapError` and `mapError` without changing the pattern here at a larger scale. Since there are other failable steps in the process, those errors too have to be transforms into a `SignalProducer(value: false)`. Also, it seems DRYer to have a pattern of flowing errors through the observer chain and then handling the errors once per step, instead of introducing the pattern that errors should be handled as locally as possible. After all, none of the framework copying steps in here handle their errors locally, but instead pass the errors through the chain to be handled by the aforementioned `flatMapError` step to produce a `SignalProducer(value: false)`.

mdiep replied:
In that case, we should make this entire function have `NoError` for the error type (since we're saying it can't error). We can still get rid of `CarthageError.internalError`—we can wrap errors in `AnyError`, send the project event, and return false (with the `NoError` type).

scottrhoyt replied:
Sounds good.

mdiep commented on Source/CarthageKit/Xcode.swift:
```diff
@@ -16,6 +16,74 @@ import XCDBLD
+}
+
+/// Parses output of `swift --version` for the version string.
+internal func parseSwiftVersionCommand(output: String?) -> String? {
```
This can be `private`.

scottrhoyt replied:
Will do.

mdiep commented on Source/CarthageKit/Xcode.swift:
```diff
@@ -16,6 +16,74 @@ import XCDBLD
+			return swiftVersion == frameworkVersion
+				? SignalProducer(value: frameworkURL)
+				: SignalProducer(error: .incompatibleFrameworkSwiftVersions(local: swiftVersion, framework: frameworkVersion))
+	}
```
This could use `attemptMap` instead of `flatMap`, which would be a bit simpler.

scottrhoyt replied:
I will take a look at that.

mdiep commented:
Unfortunately, this is going to need some conflicts resolved since #1760 was merged.

For now, I think it's enough to only apply this logic to ordinary binary releases. No need to do it for binary-only releases.

If we can resolve the conflicts and clean up the error case above, this should be all set.

scottrhoyt commented:
Sounds good, @mdiep. Do we need to add some documentation around this check in README?

mdiep commented:
Some documentation would definitely be great!

scottrhoyt commented:
Ok, I refactored that error handling. I had to extract a separate function to avoid a complex expression swift compiler error. Let me know what you think of this error flow, @mdiep, before I dive into conflict resolution.

dcaunt commented:
Just ran this on a couple of big projects here and it looks great ✨

mdiep commented:
That error handling looks good to me. Thanks!

scottrhoyt commented:
@mdiep So, I'm working on the merge conflicts right now. Combining my code with #1760 causes that `AnyError` to leak out of the install binaries step and cause many steps to either emit `AnyError`s themselves or do a lot of mapping to `AnyError`. It gets quite messy.

As I see it there are few options:

1. Create a different code path for installing binary-only dependencies vs. binaries from dependencies that can also be compiled. This gets pretty WET but allows us to contain the `AnyError`.
2. Convert many of the steps to emit `AnyError` instead of `CarthageError`. This would be a big refactor, but it would reduce the many occasions where we need to `mapError` and is probably necessary if the long term goal is to breakup `CarthageError` and not use a `CarthageError.internalError`.
3. Revert back to `CarthageError.internalError` implementation. This is quite a bit cleaner than `AnyError` and offers a pathway to breaking up `CarthageError` while still being able to flow errors easily through the steps.
4. Revert back to having the errors be `CarthageError`s. This was the cleanest in terms of `Project.swift` but contributes to the `CarthageError` problem.

As I mentioned before, if the longer term goal is to break up `CarthageError` then 2 or 3 are the best options. 2 would require a big refactor to general error handling and flow that should be done in a separate PR. So I think 3 is the way to go here.

mdiep commented:
> So I think 3 is the way to go here.

I agree. 👍 Thanks for looking into it.

scottrhoyt commented:
@mdiep done with the merge. It was much easier with the `internalError` approach. Give it a rundown and let me know if you have further feedback.

scottrhoyt commented:
@mdiep any idea on that failing test in the build. I'm not familiar with it.

mdiep commented:
The code looks good, but I've discovered an oddity.

```
$ ls
Cartfile
$ cat Cartfile 
github "antitypical/Result"
$ carthage-dev update
*** Fetching Result
*** xcodebuild output can be found in /var/folders/xj/sydx1wkd0p147x0k_7v667940000gn/T/carthage-xcodebuild.zXpwDj.log
$ ls
Cartfile          Cartfile.resolved
```

With that simple Cartfile, Carthage doesn't build Result. 😕 (`carthage-dev` is a local alias to my locally built copy of Carthage.) I don't see anything in your changes that affect this, but it also doesn't happen on master. 😩

scottrhoyt commented:
Interesting. Let me check that out, I have one idea as to where that might be coming from.

scottrhoyt commented:
Nice catch @mdiep! I could have sworn [this](https://github.com/scottrhoyt/Carthage/blob/swift_version_module_check/Source/CarthageKit/Project.swift#L582) wasn't necessary anymore, but I was wrong. Fixed now.

mdiep commented:
Thanks for all your hard work on this! This is a really great improvement, and I know Apple will be excited about it too. 🎉👏🏻

jdhealy commented on Source/CarthageKitTests/XcodeSpec.swift:
```diff
@@ -17,6 +17,8 @@ import Tentacle
 import XCDBLD
 
 class XcodeSpec: QuickSpec {
+	static let currentSwiftVersion = "3.0.2"
```
@scottrhoyt time to un-hardcode this — ❌s for https://github.com/Carthage/Carthage/pull/1866#issuecomment-290825024

scottrhoyt replied:
🤦‍♂️ will do.

jdhealy replied:
Hey, 😄 

The approach I’m currently leaning towards is having those few tests only run on CI — have an if statement check whether the environment variable for `CI` is set and not `"false"`, then put those `it` calls from Quick inside that conditional.

That condition can also include check whether some environment variable we export where we hardcode for each Xcode we use in CI.

AlexeyDemedetskiy commented:
Hi, we are experience an issue with this PR
In our framework we did disable bridging header. 
So Carthage is unable to determine Swift version by looking into it.

Any ideas (besides enabling bridging header)?

mdiep commented:
Please open a new issue.
